File: pytrim/trajectory.py
"""Simulate projectile trajectories.

Available functions:
    setup: setup module variables.
    trajectory: simulate one trajectory."""
from select_recoil import get_recoil_position
from scatter import scatter
from estop import eloss
from geometry import is_inside_target
import numpy as np
import cython

def setup():
    """Setup module variables.

    Parameters:
        None

    Returns:
        None    
    """
    global EMIN

    EMIN = 5.0  # eV


@cython.boundscheck(False)
@cython.wraparound(False)
@cython.cdivision(True)
@cython.locals(
    e=cython.double,
    dee=cython.double,
    free_path=cython.double,
    is_inside=cython.bint,
    i=cython.int
)
@cython.ccall  # Python-callable when compiled; no-op under plain Python
def trajectory(pos_init: np.ndarray, dir_init: np.ndarray, e_init: cython.double):
    """Simulate one trajectory.
    
    Parameters:
        pos_init (ndarray): initial position of the projectile (size 3)
        dir_init (ndarray): initial direction of the projectile (size 3)
        e_init (float): initial energy of the projectile (eV)

    Returns:
        ndarray: final position of the projectile (size 3)
        ndarray: final direction of the projectile (size 3)
        float: final energy of the projectile (eV)
        bool: True if projectile is stopped inside the target, 
            False otherwise
    """
    pos = pos_init.copy()
    direction = dir_init.copy()
    e = e_init
    is_inside = True

    while e > EMIN:
        free_path, p, dirp, _ = get_recoil_position(pos, direction)

        dee = eloss(e, free_path)
        e -= dee

        # pos += free_path * direction   (no NumPy broadcasting/temporaries)
        for i in range(3):
            pos[i] += free_path * direction[i]

        if not is_inside_target(pos):
            is_inside = False
            break

        direction, e, _, _ = scatter(e, direction, p, dirp)

    return pos, direction, e, is_inside

File: cytrim_cpp/trajectory.cpp
#include "trajectory.h"

void trajectory::simTrajectory(const dirVector<double, 3>& pos_init, const dirVector<double, 3>& dir_init, const double e_init, dirVector<double, 3>& pos, dirVector<double, 3>& dir, double* e, bool* is_inside)
{
    double freepath = 0;
    double p = 0;
    double dee;
    double e_recoil = 0;
    dirVector<double, 3> dirp;
    dirVector<double, 3> pos_recoil;
    dirVector<double, 3> dir_new;
    dirVector<double, 3> dir_recoil;

    pos = pos_init;
    dir = dir_init;
    *e = e_init;
    *is_inside = true;

    while(*e > EMIN){
        recoilDev.get_recoil_position(pos, dir, &freepath, &p, dirp, pos_recoil);
        dee = estopDev.eloss(*e, freepath);
        *e -= dee;

        pos = pos + (dir * freepath);
        if(geometryDev.is_inside_target(pos) == false){
            *is_inside = false;
            break;
        }

        scatterDev.doScatter(e, dir, p, dirp, dir_new, dir_recoil, &e_recoil);
        dir = dir_new;
    }
}